Examples Folder + small fix in `OPEATextEmbedder`

danielfleischer opened the pull request:
## Description

Introducing an examples folder. It includes a docker compose configuration for starting a TEI and TGI endpoints and 2 jupyter notebooks containing a summarization and RAG workflows.

Additionally, a small fix in `OPEATextEmbedder`, which is modeled after `HuggingFaceAPITextEmbedder` which accepts only a string as input. This is part of the contract of building Haystack applications and that is the expected input schema of all *Text embedders.

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [X] Others (enhancement, documentation, validation, etc.)

## Dependencies

The notebooks specify required packages for running them, i.e. `datasets, newspaper3k, lxml`.

## Tests

Built in project tests.

danielfleischer committed:
Preparing running examples

Basing on the LangChain integration project.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Running docker compose

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Global scope

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
News fetching and summary notebook

Works great. Need to update the folder's README.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Cleaning

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Text embedder should only accept a string not list[str]

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
RAG pipeline example notebook

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
pre-commit

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Examples Readme

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Update tests for TextEmbedder

It is now only accepts a string and returns a list of floats.

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
README

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Remove wrong file

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

rbrugaro commented on samples/README.md:
```diff
@@ -0,0 +1,90 @@
+
+Open the [News Summary](./news_summary.ipynb) notebook and run summarization example.
+
+Open the [Movies RAG](./movies_rag.ipynb) notebook and run RAG example.
```
there is a typo in notebook name so link is broken

danielfleischer replied:
Fixed in 46d1a1f2b9071921ca81c5f04c1c06c10dd5da33.

rbrugaro commented on samples/movies-RAG.ipynb:
I cannot see this notebook in browser
![Screenshot 2025-04-09 at 9 32 20 AM](https://github.com/user-attachments/assets/1de7a356-7e1c-4ff7-a1aa-4ccc1a432913)

danielfleischer replied:
Updated the notebook in e45a1aae8cd07c0590cbcfbafb37174baa460134.

rbrugaro commented:
@danielfleischer nice addition!
OPEA has a summarization microservice that can do truncation/refine/map_reduce based on the input context length (auto or configurable at launch time). I believe because this is not available as such in haystack adding to this initegration doesn't make sense but we can think through any other microservices from OPEA that would be nice to integrate also

rbrugaro commented on README.md:
```diff
@@ -27,12 +27,16 @@ To install the package from a pre-built wheel, run:
 
+## Examples
+
+See [Examples][./samples/] folder for a RAG and summarization notebooks, including a simple docker compose configuration for starting the OPEA backend.
```
Might be good o add some more OPEA details here? like options that can be used for the deployment for llm/embeddings and the support for GPU, Gaudi, cpu... maybe a link?

danielfleischer replied:
I added OPEA background at the top of the README.

danielfleischer committed:
Cleaning notebook metadata

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Fixed links in Readme

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer commented:
I'm realizing the audience is Haystack users wanting to learn more about OPEA, so I'll sprinkle more references, as you mentioned @rbrugaro.

danielfleischer committed:
Adding OPEA background, notebook renaming

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer commented:
> @danielfleischer nice addition! OPEA has a summarization microservice that can do truncation/refine/map_reduce based on the input context length (auto or configurable at launch time). I believe because this is not available as such in haystack adding to this initegration doesn't make sense but we can think through any other microservices from OPEA that would be nice to integrate also

I put less emphasize on the "summarization" aspect; it is a simple Haystack pipeline that contains retrieval and LLM generation, not a full fledged summarization service.

danielfleischer committed:
README

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

rbrugaro commented on samples/compose.yaml:
```diff
@@ -0,0 +1,37 @@
+
+services:
+  tei-embedding-service:
+    image: ghcr.io/huggingface/text-embeddings-inference:cpu-1.5
```
There is a more recent release: ghcr.io/huggingface/text-embeddings-inference:cpu-1.6

danielfleischer replied:
Done.

rbrugaro commented on samples/compose.yaml:
```diff
@@ -0,0 +1,37 @@
+      https_proxy: ${https_proxy}
+    command: --model-id ${EMBEDDING_MODEL_ID} --auto-truncate
+  tgi-service:
+    image: ghcr.io/huggingface/text-generation-inference:3.2.3-intel-cpu
```
also a more recent: ghcr.io/huggingface/text-generation-inference:2.4.0-intel-cpu

danielfleischer replied:
Version is newer.

rbrugaro reviewed (changes requested):
Currently, the compose files uses third party inference microservices, not OPEA. We should switch to use OPEA validated microservices that have validated and best known configs for different hardwares. 
https://github.com/opea-project/GenAIComps/blob/main/comps/llms/deployment/docker_compose/compose_text-generation.yaml
https://github.com/opea-project/GenAIComps/blob/main/comps/embeddings/deployment/docker_compose/compose.yaml
opea also has the third party base images with validated configurations like:

https://github.com/opea-project/GenAIComps/blob/main/comps/third_parties/tgi/deployment/docker_compose/compose.yaml

https://github.com/opea-project/GenAIComps/blob/main/comps/third_parties/tei/deployment/docker_compose/compose.yaml

https://github.com/opea-project/GenAIComps/blob/main/comps/third_parties/vllm/deployment/docker_compose/compose.yaml

You can see here an example how in OPEA those third party containers can be used
https://github.com/opea-project/GenAIComps/blob/main/comps/dataprep/deployment/docker_compose/compose.yaml

danielfleischer committed:
Added OPEA components to interact with TEI, TGI

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Adapted documentation for OPEA ports

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer commented:
I added OPEA embedder and text-generator, using the TEI and TGI endpoint, respectively. Verified everything works.

rbrugaro commented:
Thanks @danielfleischer for the OPEA microservice additions! The CI tests are still using the third party images. One option is to change the start_service function to call the docker compose compose.yaml that you have.

danielfleischer committed:
Updates from main

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

    CI Tests (#3)

    - Small change in an embedder class to accept only `str` and not a list of strings, based on compatibility with Haystack components, appearing also in #2.
    - Fix tests to accommodate this change.
    - Integration test script fixes.
    - CI workflow containing both unittest and integration test.

    Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Issue fixed in upstream

See https://github.com/opea-project/GenAIComps/pull/1582

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Testing CI with full compose file

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Dedicated compose file for CI; cleaning

Signed-off-by: Daniel Fleischer <daniel.fleischer@intel.com>

danielfleischer committed:
Merge branch 'main' into example

danielfleischer commented:
@rbrugaro CI tests now use OPEA images (in addition to TEI, TGI backend) in same manner as used for running the examples.
